### PR TITLE
[LOGS]: Added the possibility to dualship logs

### PR DESCRIPTION
**Pull request by ajacquemot**
### What does this PR do?

Changed the sender to always send logs to a main endpoint, keep retrying until logs are effectively sent and added additional endpoints to send logs with a best effort strategy, try once and forget.

### Motivation

Internal use only.

### Additional Notes

- [x] Write unit-tests
- [x] Add comments

### CODE REVIEW

File: pkg/logs/config/endpoints.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+package config
+
+type Endpoint struct {
+	APIKey string `mapstructure:"api_key"`
+	Logset string
```
Comment (mnshdw) [inline pkg/logs/config/endpoints.go]: I wish we could avoid this, I think we can live without it now that we have index filters.

**Review by mnshdw — DISMISSED** [review]
Nice change. I think we could strengthen the naming by having Clients +  Client or Destinations + Destination but up to you. The only complaint that I have is `logset`, can we avoid using it in new code ?

File: pkg/logs/config/endpoints.go (hunk 1 above)
Comment (tmichelet) [reply pkg/logs/config/endpoints.go]: +1 but it'll be easy to drop, and I don't think today we're fully ready

File: pkg/logs/config/endpoints.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	Host   string
+	Port   int
+}
+
+type MainEndpoint struct {
+	Endpoint
+	UseSSL       bool
```
[the remark this answers is not in the document]
Comment (tmichelet) [reply pkg/logs/config/endpoints.go]: I'll let @ajacquemot answer but things to keep in mind are that we don't want to fully support this (today) so we try to keep it simple to fit our use case

File: pkg/logs/sender/destinations.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+package sender
+
+type Destinations struct {
+	Main       *Client
+	Additonals []*Client
```
[the remark this answers is not in the document]
Comment (ajacquemot) [reply pkg/logs/sender/destinations.go]: https://en.wiktionary.org/wiki/additionals :)

File: pkg/logs/config/endpoints.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajacquemot) [reply pkg/logs/config/endpoints.go]: The idea here is to send logs to multiple datadog accounts, so we should transport them the same way.
For the proxy address, the main use case is an organisation that wants to have a look at what information is leaking out so they just need one stream of logs I think.

File: pkg/logs/pipeline/pipeline.go [hunk 4]
```diff
@@ -20,24 +20,37 @@ type Pipeline struct {
...
 // NewPipeline returns a new Pipeline
-func NewPipeline(connManager *sender.ConnectionManager, outputChan chan message.Message) *Pipeline {
-
+func NewPipeline(outputChan chan message.Message, endpoints *config.Endpoints) *Pipeline {
 	useProto := config.LogsAgent.GetBool("logs_config.dev_mode_use_proto")
 
+	// initialize the main destination
+	main := sender.NewClient(
+		sender.NewAPIKeyPrefixer(endpoints.Main.APIKey, endpoints.Main.Logset),
+		sender.NewDelimiter(useProto),
```
[the remark this answers is not in the document]
Comment (ajacquemot) [reply pkg/logs/pipeline/pipeline.go]: The goal is to send logs to Datadog for every endpoint, this looks like another feature.

File: pkg/logs/sender/client.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+func NewClient(prefixer Prefixer, delimiter Delimiter, connManager *ConnectionManager) *Client {
+	return &Client{
+		prefixer:    prefixer,
+		delimiter:   delimiter,
+		connManager: connManager,
+	}
+}
+
+func (d *Client) Write(payload message.Message) error {
```
[the remark this answers is not in the document]
Comment (ajacquemot) [reply pkg/logs/sender/client.go]: good catch

File: pkg/logs/config/endpoints.go (hunk 1 above)
Comment (ajacquemot) [reply pkg/logs/config/endpoints.go]: adding a comment though

File: pkg/logs/pipeline/pipeline.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tmichelet) [reply pkg/logs/pipeline/pipeline.go]: good point for a future proof version, for today let's not be too sophisticated :)

File: pkg/logs/sender/sender.go [hunk 6]
```diff
@@ -46,34 +39,36 @@ func (s *Sender) Stop() {
...
-		if s.conn == nil {
-			s.conn = s.connManager.NewConnection() // blocks until a new conn is ready
-		}
-		frame, err := s.delimiter.delimit(payload.Content())
+		err := s.destinations.Main.Write(payload)
 		if err != nil {
-			log.Error("can't send payload: ", payload, err)
-			continue
+			switch err.(type) {
+			case *FramingError:
```
Comment (tmichelet) [inline pkg/logs/sender/sender.go]: Could you explain what we're addressing with this `FramingError`? why do we want to fix it now?

**Comment by tmichelet** [comment]
This looks great, thanks @ajacquemot ! One question, what are we catching with the new exception handling logic? And why does it need to happen in this PR?

File: pkg/logs/sender/destination.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+// Write transforms a message into a frame and sends it to a remote server,
+// returns an error if the operation failed.
+func (d *Destination) Write(payload message.Message) error {
+	if d.conn == nil {
+		d.conn = d.connManager.NewConnection()
+	}
+
+	content := d.prefixer.prefix(payload.Content())
+	frame, err := d.delimiter.delimit(content)
+	if err != nil {
```
Comment (iksaif) [inline pkg/logs/sender/destination.go]: Here we have delimiter and frame. maybe try to use the same word

File: pkg/logs/config/endpoints.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+// Copyright 2018 Datadog, Inc.
+
+package config
+
+// Endpoint holds all the organization and network parameters to send logs to Datadog.
+type Endpoint struct {
+	APIKey string `mapstructure:"api_key"`
+	Logset string // TODO: Logset is deprecated and should be removed eventually
+	Host   string
+	Port   int
```
Comment (iksaif) [inline pkg/logs/config/endpoints.go]: I think it would be easier to put SSL and Proxy here.

File: pkg/logs/pipeline/pipeline.go [hunk 9]
```diff
@@ -20,24 +20,39 @@ type Pipeline struct {
...
+	main := sender.NewDestination(
+		sender.NewAPIKeyPrefixer(endpoints.Main.APIKey, endpoints.Main.Logset),
+		sender.NewDelimiter(useProto),
+		sender.NewConnectionManager(endpoints.Main.Host, endpoints.Main.Port, useSSL, endpoints.Main.ProxyAddress),
+	)
+
+	// initialize the additional destinations
+	var additionals []*sender.Destination
+	for _, endpoint := range endpoints.Additionals {
+		additionals = append(additionals, sender.NewDestination(
```
Comment (iksaif) [inline pkg/logs/pipeline/pipeline.go]: looks like the `NewDestination()` from and endpoint could be factorized (it looks a lot like the one above)

File: pkg/logs/sender/connection_manager.go [hunk 10]
```diff
@@ -26,16 +25,20 @@ const (
 
 // A ConnectionManager manages connections
 type ConnectionManager struct {
-	serverConfig *config.ServerConfig
+	host         string
```
Comment (iksaif) [inline pkg/logs/sender/connection_manager.go]: Looks like a `ConnectionManager` could contain or extend and `EndPoint`

**Review by iksaif — CHANGES_REQUESTED** [review]

File: pkg/logs/sender/connection_manager.go [hunk 11]
```diff
@@ -26,16 +25,20 @@ const (
...
+	useSSL       bool
 	proxyAddress string
 	mutex        sync.Mutex
 	firstConn    sync.Once
 }
 
 // NewConnectionManager returns an initialized ConnectionManager
-func NewConnectionManager(serverConfig *config.ServerConfig, proxyAddress string) *ConnectionManager {
+func NewConnectionManager(host string, port int, useSSL bool, proxyAddress string) *ConnectionManager {
 	return &ConnectionManager{
```
[the remark this answers is not in the document]
Comment (ajacquemot) [reply pkg/logs/sender/connection_manager.go]: 👍

File: pkg/logs/sender/destination.go (hunk 7 above)
Comment (ajacquemot) [reply pkg/logs/sender/destination.go]: Agreed we could merge both the prefixer and the delimeter. I'll open a PR right away after merging this one

File: pkg/logs/config/config.go [hunk 12]
```diff
@@ -63,23 +71,30 @@ func BuildServerConfig() (*ServerConfig, error) {
...
 			return nil, fmt.Errorf("could not parse logs_dd_url port: %v", err)
 		}
-		return NewServerConfig(
-			host,
-			port,
-			!LogsAgent.GetBool("logs_config.logs_no_ssl"),
-		), nil
+		main.Host = host
+		main.Port = port
+		useSSL = !LogsAgent.GetBool("logs_config.logs_no_ssl")
```
Comment (iksaif) [inline pkg/logs/config/config.go]: I would set UseProto here too, it's very similar to SSL.

File: pkg/logs/sender/destination.go (hunk 7 above)
Comment (iksaif) [reply pkg/logs/sender/destination.go]: 👍

**Review by iksaif — CHANGES_REQUESTED** [review]
Minor comments + we need a unit test (I would feel safer if we have one that really opens a connection like : https://github.com/DataDog/datadog-agent/pull/2373/files#diff-05fd3fce70913c5d7acb557b89527d12 )

File: pkg/logs/config/config.go (hunk 12 above)
Comment (ajacquemot) [reply pkg/logs/config/config.go]: 👍

File: pkg/logs/sender/prefixer_test.go [hunk 13]
```diff
@@ -3,7 +3,7 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2018 Datadog, Inc.
 
-package processor
+package sender
```
Comment (iksaif) [inline pkg/logs/sender/prefixer_test.go]: typo ?

**Review by iksaif — CHANGES_REQUESTED** [review]

**Review by iksaif — DISMISSED** [review]

**Comment by iksaif** [comment]
Approved, we will improve this and add more tests once merged with my PR. Wait for confirmation from somebody else.

File: pkg/logs/sender/destinations.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (achntrl) [reply pkg/logs/sender/destinations.go]: I meant `Additonals` -> `AdditIonals`, there is a missing `i`

**Review by mnshdw — DISMISSED** [review]

**Review by mnshdw — APPROVED** [review]

**Comment by remh** [comment]
That's great! Thank you!

**Pull request merged**